docs/modules/ROOT/pages/explanation/decisions/billing-scheduling-monitoring.adoc

```diff
@@ -0,0 +1,133 @@
+= Scheduling, Execution and Monitoring of APPUiO Billing ETL process
+
+== Problem
+
+xref:appuio-cloud:ROOT:references/architecture/metering-data-flow-appuio-cloud.adoc[Resource Usage Reporting for APPUiO Cloud] details the ETL process used to ship resource usage data to an ERP system for the purpose of billing.
+To ensure continuous and correct reporting of resource usage, this ETL process needs to be executed regularly.
+
+.Goals
+
+* The ETL process is scheduled reliably at regular intervals.
+* The ETL process is invoked for each separate product that needs to be processed.
+* Failed ETL runs are retried.
+** If retrying does not result in success, the failure is made visible through monitoring.
+** Retrying a failed run, either automatically or manually, provides the same result independent of the time at which the run is retried.
+* It is easy to manually trigger the ETL process for arbitrary past timeframes.
+
+.Non-Goals
+
+* Optimizing the ETL process itself
+* Detecting missing usage data directly in the target ERP system
+
+== Proposals
+
+=== Option 1: Cronjobs per Product
+
+This describes the current implementation, as a baseline for comparison:
+
+A Project Syn Component produces a kube cronjob for each product that requires processing.
+The cronjobs run hourly.
+The generated jobs run the ETL process for a relative timeframe ("3 hours ago until now").
+Failed jobs are exposed to monitoring via the KubeJobFailed metric.
+Corresponding alerts are generated automatically by the component.
+
+Due to the jobs running every hour, but covering a 3-hour timeframe, brief transient errors are recovered automatically:
+if a single job run fails, but the next one succeeds, the successful job will have covered the timeframe of the failed job as well.
+
+
+.Pro
+
+* Straightforward implementation
+* Automatic recovery from transient errors if they last no longer than 2 hours
+
+.Con
+
+* Re-running a failed job does not produce expected results - it will process billing data for the current time, and not the time at which the job was supposed to run.
+* Even though the system can recover from transient errors, they still generate an alert in any case, and an investigation is required to determine whether manual intervention is necessary.
+* It is not easily possible to run the ETL process for an arbitrary timeframe.
+* Running on the order of 100 jobs each hour is taxing our infrastructure.
+
+=== Option 2: Spawner Cronjob
+
+A Project Syn Component produces a single cronjob and provides it with a list of products for which the ETL process needs to run (including associated parameters).
+The cronjob runs hourly.
+On each run, the cronjob generates a new job for each product which in turn runs the ETL process.
```

Comment:
Would it be feasible to generate the jobs in such a fashion that we can circumvent [thundering herd](https://en.wikipedia.org/wiki/Thundering_herd_problem) issues? For example by injecting configurable randomized delays (on the order of a few minutes) to smear the actual work of the jobs over 15 minutes or so.

This doesn't necessarily need to be outlined here, but we may want to consider this for the implementation.

Reply:
Absolutely, and the implementation should do that.
To a degree, the issue is tamped by the job concurrency limit on the namespace. The pods run 5 at a time (even in the current implementation, which already does spawn them all at once)